### PR TITLE
Provide changes API endpoint

### PR DESCRIPTION
**Pull request by tcitworld**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documentation | no
| Translation   | no
| Fixed tickets | #2906
| License       | MIT

Start work for the changes API

### CODE REVIEW

**Comment by j0k3r** [comment]
I targetted the 2.3 endpoint (since you set the milestone to 2.3.0).
Could you then rebase your banche?

File: src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+{
+    /** @var LoggerInterface $logger */
+    private $logger;
+
+    /** @var EntityManager $em */
+    private $em;
+
+    public function __construct(EntityManager $em, LoggerInterface $logger)
+    {
+        die;
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php]: :trollface:

File: src/Wallabag/CoreBundle/Event/EntryUpdatedEvent.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Wallabag\CoreBundle\Event;
+
+use Symfony\Component\EventDispatcher\Event;
+use Wallabag\CoreBundle\Entity\Entry;
+
+/**
+ * This event is fired as soon as an entry was saved.
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/EntryUpdatedEvent.php]: Comment needs to be updated

File: src/Wallabag/CoreBundle/Event/EntryTaggedEvent.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace Wallabag\CoreBundle\Event;
+
+use Symfony\Component\EventDispatcher\Event;
+use Wallabag\CoreBundle\Entity\Entry;
+use Wallabag\CoreBundle\Entity\Tag;
+
+/**
+ * This event is fired as soon as an entry was saved.
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/EntryTaggedEvent.php]: Comment needs to be updated

File: src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        $this->em->persist($change);
+        $this->em->flush();
+
+        $this->logger->debug('saved updated entry '.$event->getEntry()->getId().' event ');
+    }
+
+    /**
+     * @param EntryDeletedEvent $event
+     */
+    public function onEntryRemoved(EntryDeletedEvent $event)
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php]: Should be `onEntryDeleted`

File: src/Wallabag/CoreBundle/Entity/Change.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace Wallabag\CoreBundle\Entity;
+
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * Change.
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Entity/Change.php]: You should add a little description on what it does

File: src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        $this->logger->debug('saved removed entry '.$event->getEntry()->getId().' event ');
+    }
+
+    /**
+     * @param EntryTaggedEvent $event
+     */
+    public function onEntryTagged(EntryTaggedEvent $event)
+    {
+        $change = new Change(Change::CHANGED_TAG_TYPE, $event->getEntry()->getId());
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/Subscriber/ChangesSubscriber.php]: In all that class, `Change` got the entry id when it should take the entry object instead.

File: src/Wallabag/CoreBundle/Entity/Change.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        return $this->timestamp;
+    }
+
+    /**
+     * @param DateTime $timestamp
+     */
+    public function setTimestamp(DateTime $timestamp)
+    {
+        $this->timestamp = $timestamp;
+    }
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Entity/Change.php]: I think you can remove all the setters.
All values are defined when you create the object and shouldn't be modified after.

File: src/Wallabag/CoreBundle/Entity/Change.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+     * @ORM\ManyToOne(targetEntity="Wallabag\CoreBundle\Entity\Entry", inversedBy="changes")
+     */
+    private $entry;
+
+    /**
+     * @var \DateTime
+     *
+     * @ORM\Column(name="timestamp", type="datetime")
+     */
+    private $timestamp;
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Entity/Change.php]: Maybe better to name it `created_at`

**Comment by nicosomb** [comment]
@tcitworld why did you create an event for article deletion?

**Comment by tcitworld** [comment]
To listen to it.

**Comment by nicosomb** [comment]
I added a relation between Change and Entry, so each time an entry is deleted, the changes are deleted too. 
And if we store only the entry ID (as you did), we have a problem with SQLite :trollface: because SQLite reuses old ID 🎉

File: app/DoctrineMigrations/Version20170328185535.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                );
+                $changeTable->addColumn(
+                    'entry_id',
+                    'integer',
+                    ['notnull' => false]
+                );
+
+                $changeTable->setPrimaryKey(['id']);
+
+                $changeTable->addForeignKeyConstraint(
```
Comment (nicosomb) [inline app/DoctrineMigrations/Version20170328185535.php]: I don't understand why the foreign key is not created here.

File: src/Wallabag/CoreBundle/Event/EntryTaggedEvent.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    /** @var Entry */
+    protected $entry;
+
+    /** @var Tag[] */
+    protected $tags;
+
+    public function __construct(Entry $entry, $tags)
+    {
+        $this->entry = $entry;
+    }
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Event/EntryTaggedEvent.php]: You forgot to set `$this->tags`

File: app/DoctrineMigrations/Version20170328185535.php (hunk 9 above)
Comment (j0k3r) [inline app/DoctrineMigrations/Version20170328185535.php]: Any error? What SQL queries did you see in your log?

Comment (nicosomb) [inline app/DoctrineMigrations/Version20170328185535.php]: OK, f*ck SQLite.

File: src/Wallabag/ApiBundle/Controller/EntryRestController.php [hunk 11]
```diff
@@ -350,9 +354,6 @@ public function deleteEntriesAction(Entry $entry)
         $em->remove($entry);
         $em->flush();
 
-        // entry deleted, dispatch event about it!
-        $this->get('event_dispatcher')->dispatch(EntryDeletedEvent::NAME, new EntryDeletedEvent($entry));
```
Comment (tcitworld) [inline src/Wallabag/ApiBundle/Controller/EntryRestController.php]: Why is this removed ?

File: src/Wallabag/CoreBundle/Controller/EntryController.php [hunk 12]
```diff
@@ -481,9 +486,6 @@ public function deleteEntryAction(Request $request, Entry $entry)
             UrlGeneratorInterface::ABSOLUTE_PATH
         );
 
-        // entry deleted, dispatch event about it!
-        $this->get('event_dispatcher')->dispatch(EntryDeletedEvent::NAME, new EntryDeletedEvent($entry));
```
Comment (tcitworld) [inline src/Wallabag/CoreBundle/Controller/EntryController.php]: Same here.

**Comment by tcitworld** [comment]
We also need events for Tags and Annotations changes.

**Comment by tcitworld** [comment]
If you agree with the statement above, let's remove the `entry` field from the Change entity and use `entity` instead. @wallabag/core

**Comment by nicosomb** [comment]
@tcitworld if we rename `entry` into `entity`, we need to remove the relation, right? And only store the ID, as you did? 
@tcitworld for the deletion event, see https://github.com/wallabag/wallabag/issues/2906#issuecomment-290071286

**Comment by tcitworld** [comment]
I was thinking of using https://www.w3.org/TR/activitypub (and provide a similar api than the Nextcloud Activity app [provides](https://github.com/nextcloud/activity/blob/master/docs/endpoint-v2.md)).

**Comment by tcitworld** [comment]
Sadly, I don't think this will be in 2.3.
